### PR TITLE
Support right outer join with other conditions

### PR DESCRIPTION
**Pull request by yibin87**
### What problem does this PR solve?

Issue Number: ref #7280 

Problem Summary:

### What is changed and how it works?
Add support for RightOuter kind join with other conditions, and do a little refact:
1. Add a new "RowFlaggedHashMapAdder" for RightSemi/RightAnti/RightOuter joins with other conditions
2. Add 'row_flagged_map' template parameter to simplify some logics

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Interpreters/Join.cpp [hunk 1]
```diff
@@ -900,9 +902,24 @@ Block Join::doJoinBlockHash(ProbeProcessInfo & probe_process_info) const
...
+            block.erase(flag_mapped_entry_helper_name);
+            if (!non_equal_conditions.other_cond_name.empty())
+            {
+                block.erase(non_equal_conditions.other_cond_name);
+            }
+            if (!non_equal_conditions.other_eq_cond_from_in_name.empty())
+            {
+                block.erase(non_equal_conditions.other_eq_cond_from_in_name);
+            }
+        }
```
Comment (SeaRise) [inline dbms/src/Interpreters/Join.cpp:922]: Move to inside `if (useRowFlaggedHashMapIfHasOtherCondition(kind))`?

Comment (yibin87) [reply dbms/src/Interpreters/Join.cpp]: ok

File: dbms/src/Interpreters/JoinPartition.cpp [hunk 2]
```diff
@@ -209,7 +209,7 @@ void JoinPartition::initMap()
     if (isCrossJoin(kind))
         return;
 
-    if (isRightSemiFamily(kind))
+    if (useRowFlaggedHashMapIfHasOtherCondition(kind))
```
[the remark this answers is not in the document]
Comment (yibin87) [reply dbms/src/Interpreters/JoinPartition.cpp]: Rename to isNecessaryKindToUseRowFlaggedHashMap, hope it is more readable. Here just want to use this function to extract the kind check part.

[the remark this answers is not in the document]
Comment (SeaRise) [reply dbms/src/Interpreters/JoinPartition.cpp]: ok~

**Review by SeaRise — APPROVED** [review]
lgtm

File: dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp [hunk 3]
```diff
@@ -218,6 +218,12 @@ Block ScanHashMapAfterProbeBlockInputStream::readImpl()
             else
                 fillColumnsUsingCurrentPartition<false, false>(num_columns_left, columns_left, num_columns_right, columns_right, row_counter_column);
             break;
+        case ASTTableJoin::Kind::RightOuter:
+            if (parent.has_other_condition)
+                fillColumnsUsingCurrentPartition<true, false>(num_columns_left, columns_left, num_columns_right, columns_right, row_counter_column);
+            else
+                fillColumnsUsingCurrentPartition<false, false>(num_columns_left, columns_left, num_columns_right, columns_right, row_counter_column);
+            break;
         default:
```
Comment (windtalker) [inline dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp:227]: In which case will the code reach this `default` branch?

File: dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp [hunk 4]
```diff
@@ -218,6 +218,12 @@ Block ScanHashMapAfterProbeBlockInputStream::readImpl()
             else
                 fillColumnsUsingCurrentPartition<false, false>(num_columns_left, columns_left, num_columns_right, columns_right, row_counter_column);
             break;
+        case ASTTableJoin::Kind::RightOuter:
```
Comment (windtalker) [inline dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp:221]: Looks like `case ASTTableJoin::Kind::RightOuter` and `case ASTTableJoin::Kind::RightAnti` is the same?

File: dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp (hunk 3 above)
Comment (yibin87) [reply dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp:227]: full join cases? Tidb join cases seems not able to reach here

File: dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp (hunk 4 above)
Comment (yibin87) [reply dbms/src/DataStreams/ScanHashMapAfterProbeBlockInputStream.cpp:221]: Yeah, they can be merged.

File: dbms/src/Interpreters/JoinPartition.h [hunk 5]
```diff
@@ -172,7 +172,7 @@ class JoinPartition
         const JoinBuildInfo & join_build_info,
         ProbeProcessInfo & probe_process_info,
         MutableColumnPtr & record_mapped_entry_column);
-    template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS, typename Maps>
+    template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS, typename Maps, bool row_flagged_map>
```
Comment (windtalker) [inline dbms/src/Interpreters/JoinPartition.h:175]: It seems we can get `row_flagged_map` from `Maps`?

Comment (yibin87) [reply dbms/src/Interpreters/JoinPartition.h:175]: It would be not that straight forward. Use row_flagged_map would be clearer.

Comment (windtalker) [reply dbms/src/Interpreters/JoinPartition.h:175]: OK

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**